### PR TITLE
first test script to check xml-input

### PR DESCRIPTION
**Pull request by mcwimm**

### CODE REVIEW

**Comment by jbathmann** [comment]
Force pushed a rebased version of current master

File: tests/CheckXMLTags/test_XMLtoProject.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+@ToDo: we need to specify required elements and tags - e.g. with DTD: each case requires DTD
+"""
+
+import unittest
+
+if __name__ == '__main__' and __package__ is None:
+    from os import sys, path
+    parent = (path.split(path.dirname(path.abspath(__file__))))[0]
+    parent = path.split(parent)[0]
+    sys.path.append(parent)
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_XMLtoProject.py]: I tried to run this script with pytest. In order to make it work, there are some changes needed.
First, the if parenthese can be removed here, cause the path always needs to be appended, if we want to run the test automatically from the main folder..

File: tests/CheckXMLTags/test_XMLtoProject.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    ## check if error is raised when file does not exist or file is not a file
+    def test_file_exists(self):
+        with self.assertRaisesRegex(OSError,
+                                    'File is not a file or does not exist'):
+            XMLtoProject(xml_project_file="tests/CheckXMLTags/notExistingFile")
+
+    ## check if error is raised when project file has no ".xml" ending
+    def test_file_extension(self):
+        with self.assertRaisesRegex(ValueError, "File is not an xml file"):
+            XMLtoProject(xml_project_file='error.txt')
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_XMLtoProject.py]: Here, we need to give the absolute path.
predefine sth like:
`pfad = path.dirname(path.abspath(__file__))`
change L 43 :

>         with self.assertRaisesRegex(ValueError, "File is not an xml file"):
            XMLtoProject(xml_project_file=pfad+'/error.txt')

File: tests/CheckXMLTags/test_XMLtoProject.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+Actions that require an error message to appear
+- no xml-input file/ xml-file not found
+- type: xml - @Jasper: enough?
+- empty xml-file - @Jasper: necessary?
+- required xml-tree-elements are missing (e.g. tree_dynamics)
+- required subelements are missing (e.g. aboveground_competition)
+- required xml-tags are missing (e.g. type, species)
+
+@ToDo: we need to specify required elements and tags - e.g. with DTD: each case requires DTD
+"""
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_XMLtoProject.py]: This can be done when PR #9 and #10 are merged.

File: tests/CheckXMLTags/test_XMLtoProject.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    from os import sys, path
+    parent = (path.split(path.dirname(path.abspath(__file__))))[0]
+    parent = path.split(parent)[0]
+    sys.path.append(parent)
+from ProjectLib import XMLtoProject
+from lxml import etree
+
+
+class TestXMLToProject(unittest.TestCase):
+    from lxml import etree
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_XMLtoProject.py]: I think you do not need to import lxml twice. I'd suggest to import it at the begin of the file

File: tests/CheckXMLTags/test_xmlfile.xml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_xmlfile.xml]: Maybe you can rename this file and establish a new subfolder.
sth like:
tests/Benchmarks/..
Than we can state somewhere, that all project-files from the benchmarks schould be checked for consistency...

File: tests/CheckXMLTags/test_Simple_Test_Tags.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
```
Comment (jbathmann) [inline tests/CheckXMLTags/test_Simple_Test_Tags.py]: As I see it, this file can be removed? 
You can just drop the corresponding commit using
git rebase -i master

File: main.py [hunk 7]
```diff
@@ -4,6 +4,9 @@
 from ProjectLib import XMLtoProject
 from TimeLoopLib import TreeDynamicTimeStepping
 
-prj = XMLtoProject(xml_project_file="ProjectLib/testproject.xml")
+# prj = XMLtoProject(xml_tree=
+#                    "tree_dynamics")
+
+prj = XMLtoProject(xml_project_file="tests/CheckXMLTags/test_xmlfile.xml")
```
Comment (jbathmann) [inline main.py]: Please drop those changes

File: ProjectLib/XMLtoProject.py [hunk 8]
```diff
@@ -15,20 +15,38 @@
...
+                self.prjfile = args["xml_project_file"]
+                self.readProjectFile()
+                self.addTreeDynamicConcepts()
+                self.addInitialPopulation()
+                self.addTreeTimeLoop()
+
+                self.argsToProject()
+
+        elif "xml_tree" in args:
+            tree = args["xml_tree"]
```
Comment (jbathmann) [inline ProjectLib/XMLtoProject.py]: Stimmt das so? wird gecheckt, ob der tree alle benötigten tags hat?
die funktion self.argsToProject() muss erst nach diesem statement aufgerufen werden, oder=

File: ProjectLib/Project.py [hunk 9]
```diff
@@ -21,6 +21,9 @@ def argsToProject(self):
         self.iniPopulation()
         self.iniTreeTimeLoop()
 
+    def setProjectArguments(self, args):
+        self.args = args
```
Comment (jbathmann) [inline ProjectLib/Project.py]: Puh, This is confusing indeed. Can you explain the project creation sheme from an etree-object to me please?

File: .gitignore [hunk 10]
```diff
@@ -106,3 +106,5 @@ Documentation/html/
 Documentation/latex/
 html
 latex
+
+.idea
```
Comment (jbathmann) [inline .gitignore]: This folder is created by unittest?

File: tests/CheckXMLTags/test_XMLtoProject.py (hunk 4 above)
Comment (mcwimm) [reply tests/CheckXMLTags/test_XMLtoProject.py]: changed

File: tests/CheckXMLTags/test_xmlfile.xml (hunk 5 above)
Comment (mcwimm) [reply tests/CheckXMLTags/test_xmlfile.xml]: Would you consider this test-file as a benchmark? I would create 2 folders: one with general testing script (like this xml-test) and another with benchmarks (use a specific input to create a specific output, e.g. test funtionality of simpleBettina).

File: .gitignore (hunk 10 above)
Comment (mcwimm) [reply .gitignore]: no, my IDE creates .idea.

File: tests/CheckXMLTags/test_xmlfile.xml (hunk 5 above)
Comment (jbathmann) [reply tests/CheckXMLTags/test_xmlfile.xml]: Well, it's not a benchmark yet, but can simply become a simple one. Having this script with all the dummy functions we know exactly which output we would expect.

File: .gitignore (hunk 10 above)
Comment (jbathmann) [reply .gitignore]: :)

File: tests/CheckXMLTags/test_XMLtoProject.py (hunk 2 above)
Comment (mcwimm) [reply tests/CheckXMLTags/test_XMLtoProject.py]: changed

File: tests/CheckXMLTags/test_Simple_Test_Tags.py (hunk 6 above)
Comment (mcwimm) [reply tests/CheckXMLTags/test_Simple_Test_Tags.py]: I'll try

File: main.py (hunk 7 above)
Comment (mcwimm) [reply main.py]: s.o.

File: tests/CheckXMLTags/test_XMLtoProject.py (hunk 1 above)
Comment (mcwimm) [reply tests/CheckXMLTags/test_XMLtoProject.py]: resolved

File: ProjectLib/XMLtoProject.py [hunk 11]
```diff
@@ -33,6 +33,8 @@ def __init__(self, **args):
             else:
                 self.prjfile = args["xml_project_file"]
                 self.readProjectFile()
+                self.checkRequiredElements()
```
Comment (jbathmann) [inline ProjectLib/XMLtoProject.py]: this is great!

File: ProjectLib/XMLtoProject.py [hunk 12]
```diff
@@ -42,6 +44,41 @@ def __init__(self, **args):
...
+                if L2 is None:
+                    raise KeyError("Element '" + l2 + "' in '" +
+                                   l1 + "' is missing!")
+
+                if l1 in level3:
+                    for l3 in level3[l1]:
+                        L3 = L2.find(l3)
+                        if L3 is None:
+                            raise KeyError("Element '" + l3 + "' in '" +
+                                   l1 + "/" + l2 + "' is missing!")
```
Comment (jbathmann) [inline ProjectLib/XMLtoProject.py]: Here and for the other key errors:
Maybe you can make the error message more instuctive for the user. 
Sth like: "Required element 'blabla' in project configuration is missing. Please check your manga-project file.."

**Comment by jbathmann** [comment]
Just rebased and force-pushed. Please be careful before resetting old branches. The rebase was quite messy - dunno whether i did sth wrong

**Comment by jbathmann** [comment]
This work is continued with PR #37

**Pull request closed without merging**